### PR TITLE
refactor: throw on not found resources by default

### PR DESCRIPTION
**Pull request by tpluscode**
This changes the default of `ResourceStore` to throw when a resource is not found. A second parameter can be added where an `undefined` is acceptable. Types adjust automatically which removes the need to the repetitive `if (!resource) throw new Error()` boilerplate

### CODE REVIEW

**Review by martinmaillard — APPROVED** [review]

**Review by lucafurrer — APPROVED** [review]
:+1:

**Review by lucafurrer — APPROVED** [review]

File: apis/core/lib/ResourceStore.ts [hunk 1]
```diff
@@ -110,31 +116,40 @@ export default class implements ResourceStore {
...
-    if (!this.__resources.has(term)) {
-      const resource = await this.__storage.loadResource(term)
-      if (resource) {
-        this.__resources.set(term, resource)
+
+    if (!resource) {
+      if (opts?.allowMissing) {
+        return undefined as any
       }
+      throw new NotFoundError(term)
```
Comment (martinmaillard) [inline apis/core/lib/ResourceStore.ts:145]: If I'm not mistaken, `NotFoundError` results in a 404 response from the API. I think in a lot of cases that's not what we want when loading a resource from the store. 404 should only happen when the resource at the requested URI doesn't exist, but not when another resource is not found while processing the request.

Comment (tpluscode) [reply apis/core/lib/ResourceStore.ts:145]: Fair point 👌

Comment (martinmaillard) [reply apis/core/lib/ResourceStore.ts]: I see your fix, but now how about all the cases that *should* be 404? :D

Comment (tpluscode) [reply apis/core/lib/ResourceStore.ts]: hydra-box will do that earlier, before it gets to our handlers

Comment (martinmaillard) [reply apis/core/lib/ResourceStore.ts]: Oh, alright then :)

**Review by martinmaillard — APPROVED** [review]

**Pull request merged**